text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,236 @@
+But this shifts the burden to individual packages, and means we'd still need to vendor source of any crate moved to crates.io in the compiler source registry for packages that didn't make the switch.
+
+More interesting would be to change Cargo's *default* behavior to check both the compiler-specific sources (compiler source registry and sysroot binary mock registry) and crates.io.
+This would allow standard crates to seamlessly migrate to crates.io without extra work per package.
+This could be either be done where crates.io overrides the compiler-specific sources, or the compiler specific sources override crates.io.
+We don't want to commit to either variant in this RFC, however, so we instead want to keep all 3 options open (no fallback, crates.io over compiler-specific, compiler-specific over crates.io).
+To achieve this, we want to keep the sysroot/compiler source registry and crates.io disjoint: no package should be contained in both sources.
+That way unioning them together with either priority (the fallback scheme effectively crates a union source) has the same effect.
+
+The easiest way to achieve this is to make sure that standard library crates use names reserved on crates.io.
+We don't want to bake crates.io policy into Cargo however, so instead of absolutely prohibiting stdlib deps with non-reserved names, crates.io will just lint packages being uploaded.
+Also, care will be taken so that any stdlib crate that is stabalized must use a reserved name or already be published on crates.io.
+That still doesn't protect unpublished packages using unstable stdlib crates without reserved names from breakage, but due to their use of unstable interfaces we have no obligation to keep them working.
+Also, once we have an option to explicitly provide the source for stdlib deps, they can force the behavior they want.
+This seems good enough.
+
+
+# Drawbacks
+
+ - The mock registry for sysroot binaries is a disgusting hack.
+
+ - Only some crates in the rust repo (at least `core`, `alloc` and `collections`) can properly be built just based upon their `Cargo.toml`.
+   However it's precisely only these "foundational" crates that will be of interest to freestanding developers.
+   Hosted developers can likely get pre-built binaries for the platform they need with `rustup`, just as they do today.
+
+ - Compilers could provide crates in their sysroot that don't match the Rust specification, and Cargo would be none the wiser.
+   (Technically, this problem already exists with falling back on the sysroot binaries, but users will probably expect better when they can specify standard library dependencies explicitly.)
+   Since the *interface* of the stdlib is specified, it would be neat if we could put a big crate type/interface on crates.io, which compiler implementations would need to match.
```

Comment:
Maybe this should emphasize that the problem is with stable crates specifically. Compilers providing unstable crates is necessary.